### PR TITLE
Preview order functionality complete

### PR DESCRIPTION
**Pull request by PaulaRudy**
Preview order now reflects position in the pipeline or in the list of sources. Source previews are always displayed at the beginning. Previews now react to steps being moved in the pipeline.

### CODE REVIEW

**Comment by PaulaRudy** [comment]
So at this point functionality is complete. I've tested it by hand and everything works. I would like to pull out some methods from the onPreviewOrderChanged method and add some tests, but as of this moment everything works fine.

**Comment by JLLeitschuh** [comment]
Uhhhhh. Yea, you added all of the generated files to the repository.
Can you undo that please?

**Comment by PaulaRudy** [comment]
All fixed! Thanks for catching that.

File: ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java [hunk 1]
```diff
@@ -154,10 +154,12 @@ private OutputSocketView findOutputSocketView(OutputSocket socket) {
     }
 
     /**
+     * This function is public because it is used by {@link edu.wpi.grip.ui.preview.PreviewsView#onSocketPreviewChanged}
+     * to order the list of previews.
+     *
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java]: This comment isn't really necessary and, as @ThomasJClark said before, Intellij does this for you by default.

File: ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java [hunk 2]
```diff
@@ -168,9 +170,12 @@ private SourceView findSourceView(Source source) {
...
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java]: Same as above.

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java [hunk 3]
```diff
@@ -48,24 +58,196 @@ public PreviewsView(EventBus eventBus) {
...
+                        boolean zeroReached = false;//We will set this to true if we reach the beginning of the list of previews (there are no source previews open)
+                        while ((!zeroReached) &&
+                                ((nextSocketInDirection.getStep().isPresent())
+                                        && (nextSocketInDirection.getStep().get() == this.previewedSockets.get(newLocation).getStep().get()))) { //While we haven't reached the beginning of the list of previews, the socket at this location is a socket from a step, and it is the SAME step as the step of the socket at the new location...
+                            count++;
+                            if ((newLocation - count) > 0)//If we haven't reached the beginning of the list of open previews...
+                                nextSocketInDirection = this.previewedSockets.get(newLocation - count);//Grab the next previewed socket to examine in the direction we are moving
+                            else
+                                zeroReached = true;//Mark that we've reached the beginning of the list of previews so we know to stop looking for more
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: So as a general practice always wrap your if statement code in blocks `{ ... }` OR put it on the same line
`else zeroReached = true;`

I enforce this because merging code can create bugs that are really hard to spot.
See the major apple security bug that was caused by having two gotos after an if statement: https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/

File: ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java (hunk 2 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java]: @JLLeitschuh :+1:

File: ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java (hunk 1 above)
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java]: @JLLeitschuh :+1:

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java [hunk 4]
```diff
@@ -48,24 +58,196 @@ public PreviewsView(EventBus eventBus) {
     }
 
     @Subscribe
+    /**
+     * This function is called when a step moves in the pipeline to adjust the positions of any open previews it has
+     * to reflect the new order of the pipeline.
+     */
+    public synchronized void onPreviewOrderChanged(StepMovedEvent event) {
```
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: So, just curious, but is there a reason why you didn't use some pre-defined sorting function?  Seems like it would be nice to not have all of this manual keeping track of positions and stacks and stuff.

    ObservableList<SocketPreviewView<?>> previews = FXCollections.observableArrayList(
            (Collection) previewBox.getChildren());

    Collections.sort(previews, (a, b) -> {
        if (...) {
            return -1; // a goes before b
        } else {
            return -1; // b goes before a
        }
    });

    previewBox.getChildren().setAll(previews);

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java (hunk 3 above)
Comment (PaulaRudy) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: Fair enough :)

File: ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java (hunk 1 above)
Comment (PaulaRudy) [inline ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java]: Ok. Will fix.

File: ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java (hunk 2 above)
Comment (PaulaRudy) [inline ui/src/main/java/edu/wpi/grip/ui/pipeline/PipelineView.java]: Ok. Will fix.

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java (hunk 4 above)
Comment (PaulaRudy) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: To be honest, I'd rather just leave it as is and come back to this later- at this point, logging seems more important to me. Is this something that you think is critical before this can be merged? I can always come back to this later. But I do think you're right about this method being cleaner. I avoided this in this instance because I forgot about lambdas (I learned java without them). At the very least, next time I will be sure to use this method. Thank you for bringing it up :)

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java [hunk 5]
```diff
@@ -28,12 +35,15 @@
 
     private final EventBus eventBus;
     private final List<OutputSocket<?>> previewedSockets;
+    private final PipelineView pipeline;//This is used to determine the order the previews are shown in.
 
-    public PreviewsView(EventBus eventBus) {
+    public PreviewsView(EventBus eventBus, PipelineView pipeline) {
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: :+1: Do this please.

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java (hunk 4 above)
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: I'm going to agree with @PaulaRudy on this one. There are more important things to be taken care of.
If we have problems in the future this can be refactored.

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java [hunk 6]
```diff
@@ -48,24 +58,203 @@ public PreviewsView(EventBus eventBus) {
     }
 
     @Subscribe
+    /**
+     * This function is called when a step moves in the pipeline to adjust the positions of any open previews it has
+     * to reflect the new order of the pipeline.
+     */
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: Can you move the comment  to above the `@Subscribe` attribute. Its the code style we've been following everywhere else and here it looks out of place.

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java [hunk 7]
```diff
@@ -48,24 +58,203 @@ public PreviewsView(EventBus eventBus) {
...
+                this.previewedSockets.add(newLocation, current);//...add it to the correct location in the list of previews open
+                this.previewBox.getChildren().add(newLocation, SocketPreviewViewFactory.createPreviewView(this.eventBus, current));//...and display it in the correct location in the list of previews open
+            }
+        });
+    }
+
+    @Subscribe
+    /**
+     * This function is called when a preview button is pushed/triggered
+     */
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: Same as above,
comments above the `@Subscribe`

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java [hunk 8]
```diff
@@ -48,24 +58,203 @@ public PreviewsView(EventBus eventBus) {
     }
 
     @Subscribe
+    /**
+     * This function is called when a step moves in the pipeline to adjust the positions of any open previews it has
+     * to reflect the new order of the pipeline.
+     */
+    public synchronized void onPreviewOrderChanged(StepMovedEvent event) {
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: I do agree with this comment. The complexity is high. If we have problems we may have to refactor it.

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java (hunk 6 above)
Comment (PaulaRudy) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: Sure!

File: ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java (hunk 5 above)
Comment (PaulaRudy) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: So this is an issue with the original code, not my changes. Is this still something you would like me to do? Shouldn't it be in a separate pull request?

Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java]: Just take care of it here. You might as well fix this while you're here.
